lib: cusparse: fix #1075 (conversion from sparse matrix to dense matrix)

frapac opened the pull request:
Fix issue #1075 
Now the following code works:
```julia
julia> CUDA.allowscalar(false)

julia> x = sprand(5,5, 0.2) |> cu
5×5 CuSparseMatrixCSC{Float32} with 8 stored entries:
 0.352324   ⋅         ⋅         ⋅        ⋅
  ⋅         ⋅         ⋅         ⋅        ⋅
  ⋅        0.781994   ⋅         ⋅        ⋅
 0.927318   ⋅         ⋅        0.58433  0.922552
 0.370246   ⋅        0.593702   ⋅       0.0792545

julia> x |> Array
5×5 Matrix{Float32}:
 0.352324  0.0       0.0       0.0      0.0
 0.0       0.0       0.0       0.0      0.0
 0.0       0.781994  0.0       0.0      0.0
 0.927318  0.0       0.0       0.58433  0.922552
 0.370246  0.0       0.593702  0.0      0.0792545

julia> x |> CuArray
5×5 CuArray{Float32, 2, CUDA.Mem.DeviceBuffer}:
 0.352324  0.0       0.0       0.0      0.0
 0.0       0.0       0.0       0.0      0.0
 0.0       0.781994  0.0       0.0      0.0
 0.927318  0.0       0.0       0.58433  0.922552
 0.370246  0.0       0.593702  0.0      0.0792545
```

frapac committed:
cusparse: fix conversion from sparse matrix to dense matrix

maleadt commented on lib/cusparse/conversions.jl:
```diff
@@ -326,6 +326,8 @@ for (elty, welty) in ((:Float16, :Float32),
     end
 end
 
+Base.Array{T, 2}(A::AbstractCuSparseMatrix{T}) where T = Base.Array{T, 2}(CuMatrix{T}(A))
```
Would it not be better to implement the underlying `copyto!` call?

maleadt commented:
Thanks!

CI failures look related.

frapac commented:
Oups, indeed. I had some problems to run the CI locally, with #1077. I will update the deps according to your explanation in that issue and re-check everything.

frapac committed:
fix mv! and mm! tests on PR#1078

- revert `CuSparseMatrixDescriptor(::CuSparseMatrixCSC)` to origin
- add a new function `create_csc_descriptor`
- move new conversion tests in `tests/cusparse.jl`

frapac commented:
The problem was the following: 
- `mv!` and `mm!` do not support `CSC` matrix by default (see [doc](https://docs.nvidia.com/cuda/cusparse/index.html#cusparse-generic-function-spmv)). The workaround in CUDA.jl is to use instead a CSR description to do the matrix multiplication (see constructor of `CuSparseMatrixDescriptor`)
- however, if we use the CSR description in the conversion, then `d_A |> CuArray` does not return the solution expected, but its transpose. We have to use a CSC description there. 
- To fix that, the latest commit introduces a new function `create_csc_description` to create manually the correct description for CSC description. We keep `CuSparseMatrixDescriptor` as is to avoid breaking the `mv!` and `mm!` functions.

maleadt commented:
Looks good, thanks. Maybe add some comments though, e.g. why there's `create_csc_description` and not just the constructor (i.e. what you explained in the comment above).

frapac committed:
add comments for create_csc_descriptor

maleadt commented:
Continuing in https://github.com/JuliaGPU/CUDA.jl/pull/1083 since I couldn't push to your fork. It's better if you check the `Allow maintainers to edit this PR` checkbox.

frapac commented:
I didn't know that. I will do that in the future. Continuing in #1078 looks good to me.
